src/main/java/org/truffleruby/extra/RubyConcurrentHashMap.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (c) 2015, 2020 Oracle and/or its affiliates. All rights reserved. This
+ * code is released under a tri EPL/GPL/LGPL license. You can use it,
+ * redistribute it and/or modify it under the terms of the:
+ *
+ * Eclipse Public License version 2.0, or
+ * GNU General Public License version 2, or
+ * GNU Lesser General Public License version 2.1.
+ */
+package org.truffleruby.extra;
+
+import com.oracle.truffle.api.object.Shape;
+import org.truffleruby.RubyContext;
+import org.truffleruby.core.basicobject.RubyBasicObject;
+import org.truffleruby.core.klass.RubyClass;
+import org.truffleruby.language.RubyDynamicObject;
+
+import java.util.concurrent.ConcurrentHashMap;
+
+public class RubyConcurrentHashMap extends RubyDynamicObject {
+
+   public static class Key {
+
+       public final Object key;
+
+       public Key(Object key) {
+           this.key = key;
+       }
+
+       @Override
+       public int hashCode() {
+           final Object code = RubyContext.send(key, "hash");
+           if (code instanceof Integer) {
+               return (int) code;
+           } else if (code instanceof Long) {
+               return (int) (long) code;
+           } else {
+               throw new UnsupportedOperationException(code.getClass().getName());
+           }
```

Comment:
We should be storing the hash in the object - remember that discussion? So hash once when inserting and then store the hash.

Reply:
Right, so let's have a `final int hashCode` field, and in each node that creates a `Key`, compute the hash code first with `ToHashByHashCode`, and pass it to the Key. Then hashCode() just returns the field.

Reply:
Done.